refactor: Add @packages/errors, type safety & consistency for errors

tgriesser opened the pull request:
Part of the work for #19872 involves getting better structure around the errors. 

- Adds `@packages/errors` & `errTemplate` for a consistent way of structuring our errors. Any formatting changes should be added in the `errTemplate`.

- Added [documentation](https://github.com/cypress-io/cypress/blob/e12e2acebc8c6d4dfefedfd7fb02841a69ee1b6a/guides/error-handling.md) around these error changes & best practices

- Adds snapshotted html for each error's console rendering to easily view all of the errors in once place and make sure any formatting changes are applied correctly.

- [x] Move from a switch statement to an object literal so we can pull out all the unique object keys.
- [x] Consistent return types from the tagged literal so we can have a consistent shape returned from each error

tgriesser committed:
chore: rename errors.js -> errors.ts

tgriesser committed:
refactor: type safety on errors

tgriesser committed:
refactor: add err_template for consistent error formatting

brian-mann commented on packages/server/__snapshots__/cypress_spec.js:
```diff
@@ -302,7 +302,8 @@ The error was: Cannot read property 'split' of undefined
 
 exports['INVALID_CONFIG_OPTION'] = `
-\`test\` is not a valid configuration option,\`foo\` is not a valid configuration option
+\`test\` is not a valid configuration option
+\`foo\` is not a valid configuration option
```
use `listItems` here - I'll take this

brian-mann commented on packages/server/lib/browsers/utils.ts:
```diff
@@ -1,10 +1,12 @@
 // @ts-ignore
 import plugins from '../plugins'
 
+const errorsChild = require('../errors-child')
```
why did we split out this single error into `errors-child` ?

tgriesser replied:
Because the child process uses these errors and right now the child process only allows for js files. Wasn't going to tackle that as part of this PR

brian-mann reviewed (changes requested):
- since this is a PR into `develop` - I'll add a corresponding issue that describes these changes
- add system tests for plugin errors - I'll add these

tgriesser committed:
fix a few system tests

brian-mann committed:
fix tests; update snapshots

tgriesser committed:
Fix types

brian-mann committed:
normalize snapshot - remove chalk ansi colors

brian-mann committed:
Merge branch 'tgriesser/chore/refactor-errors' of https://github.com/cypress-io/cypress into tgriesser/chore/refactor-errors

brian-mann committed:
more unit test fixes

brian-mann committed:
more system test fixes

brian-mann committed:
circleci build

brian-mann committed:
backtick always in stdout, fix error formatting and failing snapshots

tgriesser committed:
Merge branch 'develop' into tgriesser/chore/refactor-errors

tgriesser committed:
refactor: create @packages/errors

tgriesser committed:
fix import

brian-mann reviewed (changes requested):
It looks like it renamed `__snapshots__/errors_spec.js` to `__snapshots__/errors_spec.ts.js`

but then added a new file here https://github.com/cypress-io/cypress/pull/19952/commits/32f3e1874c5afeb4de297c7ea5be34537ba14c35#diff-7121d82c702e5f418419b336ca519cab0b98d9736769a3305498bd0391fc13d4

I think one of these files is errant

brian-mann commented on packages/errors/src/errorTypes.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+export interface ErrTemplateResult {
+  message: string
+  details?: string
```
now that we cleaned up the normalization of the properties shouldn't these 2 properties be put under like a `forTerminal()` function? this would match the graphql schema better

brian-mann replied:
Since the `message` and `details` top level are only used for the terminal - whereas the `forBrowser()` returned `message` and `details` are only used for the browser

brian-mann replied:
From slack convo with tim:

> Yeah I was thinking that might be part of the 10 PR rather than adding more surface area to this one, it would be a massive change and the 9.x app doesn't even use forBrowser right now
>
> Bc it does ansi stripping in the browser or something

brian-mann commented on packages/errors/src/error_utils.ts:
```diff
@@ -0,0 +1,89 @@
+  return err
+}
+
+export const warnIfExplicitCiBuildId = function (ciBuildId?: string | null) {
+  if (!ciBuildId) {
+    return ''
+  }
+
+  return `\
+It also looks like you also passed in an explicit --ci-build-id flag.
+
+This is only necessary if you are NOT running in one of our supported CI providers.
+
+This flag must be unique for each new run, but must also be identical for each machine you are trying to --group or run in --parallel.\
+`
+}
```
does this need to be in `error_utils` or could it go into the main `errors` file? I think it's only used from within `errors`

tgriesser replied:
Moved

brian-mann commented on packages/server/lib/errors.ts:
```diff
@@ -1,1246 +1,10 @@
+import errors from '@packages/errors'
```
did you mean to delete this file?

tgriesser replied:
No, it's meant to keep some of the existing tests & exist code in place without needing to refactor

brian-mann commented on packages/server/lib/plugins/child/task.js:
```diff
@@ -1,6 +1,6 @@
 const _ = require('lodash')
 const util = require('../util')
-const errorsChild = require('../../errors-child')
+const errorsChild = require('../../../../errors/src/errors-child')
```
`@packages/error` ?

tgriesser committed:
fix import

tgriesser committed:
fixing build / tests

tgriesser committed:
remove extraneous file

tgriesser committed:
move warnIfExplicitCiBuildId

tgriesser committed:
fix build / tests

tgriesser committed:
Fix

brian-mann commented on packages/server/test/unit/errors_spec.js:
```diff
@@ -3,6 +3,10 @@ const chalk = require('chalk')
 
 context('.logException', () => {
+  beforeEach(() => {
+    sinon.stub(console, 'log')
```
you usually want `sinon.spy(...)` here so you can still see the console.logs

brian-mann commented:
the tests are almost passing - but prior to merge we still need to do a once over review on all the errors to ensure it was all formatted correctly

tgriesser committed:
error, type fixes, documentation, standardize child process error serialization

tgriesser committed:
fix import

tgriesser committed:
build errors on install

brian-mann committed:
wrote specs generating visual images of all errors

brian-mann committed:
remove unused dep

brian-mann committed:
sanitize stack traces

brian-mann committed:
add image diffing

- if base images don't exist, create them
- if base images don't match and local, overwrite them, if in CI throw
- if base images are stale and local, delete them, if in CI throw

tgriesser committed:
type fixes, remove Bluebird, general cleanup

tgriesser committed:
TS Cleanup

tgriesser committed:
skip typecheck on tests for now

tgriesser committed:
yarn.lock

tgriesser committed:
fix @types/chai version

tgriesser committed:
fix yarn.lock

tgriesser committed:
Different version of mocha types so it isnt patched

brian-mann committed:
errors spec snapshot

brian-mann committed:
remove Courier New + MesloLGS NF font

tgriesser committed:
CI fix

tgriesser committed:
fixes

brian-mann committed:
store snapshot images in circle artifacts

brian-mann committed:
dont change artifact destination prefix

brian-mann committed:
use Courier Prime

brian-mann committed:
antialias the text

brian-mann committed:
decrease pixelmatch threshold, fail in CI only when changed pixels > 100

brian-mann committed:
increase timeout

tgriesser committed:
run unit tests w/ concurrency=1

tgriesser committed:
unique window per file

brian-mann committed:
disable app hardware acceleration + use in process gpu + single process

brian-mann committed:
do not do image diffing

- conditionally convert html to images
- store html snapshots
- do not store images in git

brian-mann committed:
store snapshot html

brian-mann committed:
Merge branch 'tgriesser/chore/refactor-errors' of https://github.com/cypress-io/cypress into tgriesser/chore/refactor-errors

brian-mann committed:
remove concurrency

brian-mann committed:
fix assertion

tgriesser committed:
fixing ci

tgriesser committed:
Link in readme

jennifer-shehane commented:
@tgriesser I'd like to have @chrisbreiding  review these changes since he worked in this most extensively in the past.

tgriesser committed:
pass the browsers to listItems

chrisbreiding reviewed (dismissed):
LGTM 👍

tgriesser committed:
fix: build @packages/errors in CI, defer import to prevent errors locally

tgriesser committed:
Merge branch 'develop' into tgriesser/chore/refactor-errors

* develop:
  chore: fix cypress npm package artifact upload path (#20023)
  chore(driver): move cy.within logic into it's own file (#20036)
  chore: update automerge workflows (#19982)
  fix(selectFile): use target window's File/DataTransfer classes (#20003)
  chore: Update Chrome (stable) to 98.0.4758.80 and Chrome (beta) to 98.0.4758.80 (#19995)
  fix: Adjust ffmpeg CLI args for performance (#19983)
  build: allow unified to run cypress on Apple Silicon (arm64) (backport #19067 to 9.x) (#19968)

tgriesser committed:
fix run-if-ci.sh

brian-mann committed:
overflow: hidden, remove new Promise, add debug logging

Co-Authored-By: Tim Griesser <tgriesser@gmail.com>

brian-mann committed:
Merge branch 'develop' into tgriesser/chore/refactor-errors

brian-mann committed:
remove dead code

tgriesser committed:
Mark the .html files as generated in gitattributes

brian-mann committed:
fix running single error case, slice out more of the brittle stack

brian-mann committed:
remove additional brittle stack line

tgriesser commented:
Closing in favor of #20072
